ptbcontrib/roles/roles.py

```diff
@@ -0,0 +1,645 @@
+        child_roles (:class:`telegram.ext.Role` | set(:class:`telegram.ext.Role`)), optional):
+            Child roles of this role.
+        name (:obj:`str`, optional): A name for this role.
+    """
+
+    _DEFAULT_ADMIN_NAME: ClassVar[str] = 'ptbcontrib_roles_default_admin'
+    _admin_lock = Lock()
+    _admin_event = Event()
+    _admin: ClassVar['Role'] = None  # type: ignore[assignment]
+
+    def __init__(
+        self,
+        chat_ids: Union[int, List[int], Tuple[int, ...]] = None,
+        child_roles: Union['Role', List['Role'], Tuple['Role', ...]] = None,
+        name: str = None,
+    ) -> None:
+        super().__init__(chat_ids)
+        self._name = name
+        self._inverted = False
+        self.__lock = Lock()
+
+        self._child_roles: Set['Role'] = set()
+        self._set_child_roles(child_roles)
+
+        # We need the if clause for the init of _admin
+        if name != self._DEFAULT_ADMIN_NAME:
+            self.__init_admin()
+            self._admin_event.wait()
+            self._admin.add_child_role(self)
+
+    @staticmethod
+    def __init_admin() -> None:
+        with Role._admin_lock:
+            if Role._admin is None:
+                Role._admin = Role(name=Role._DEFAULT_ADMIN_NAME)
+                Role._admin_event.set()
+
+    def _set_custom_admin(self, new_admin: 'Role') -> None:
+        with self._admin_lock:
+            self._admin.remove_child_role(self)
+        new_admin.add_child_role(self)
+
+    @staticmethod
+    def _parse_child_role(
+        child_role: Union['Role', List['Role'], Tuple['Role', ...], None]
+    ) -> Set['Role']:
+        if child_role is None:
+            return set()
+        if isinstance(child_role, Role):
+            return {child_role}
+        return set(child_role)
+
+    def _set_child_roles(
+        self, child_role: Union['Role', List['Role'], Tuple['Role', ...], None]
+    ) -> None:
+        with self.__lock:
+            self._child_roles = self._parse_child_role(child_role)
+
+    @property
+    def child_roles(self) -> FrozenSet['Role']:
+        """Child roles of this role. This role can do anything, its child roles can do.
+        May be empty.
+
+        Returns:
+            Set(:class:`telegram.ext.Role`):
+        """
+        with self.__lock:
+            return frozenset(self._child_roles)
+
+    def __invert__(self) -> 'InvertedRole':
+        return InvertedRole(self)
+
+    def filter(  # pylint: disable=W0221
+        self,
+        update: Update,
+        target: 'Role' = None,
+        inverted: bool = False,
+    ) -> Optional[bool]:
+        # Always allow admins
+        if self is not self._admin and self._admin.filter(update):
+            return not inverted
+
+        user = update.effective_user
+        chat = update.effective_chat
+
+        # If the update has neither effective chat nor user, we don't handle it
+        if not (user or chat):
+            return False
```

Comment:
For inverted filters, this will give a false positive. Need to clarify and test inversion up to the `if inverted` part